### PR TITLE
add thanos config_timeout and config_interval args

### PR DESCRIPTION
**Pull request by viennaa**
## Description

Adding missing arguments introduced in [Thanos in September, 2022](https://github.com/thanos-io/thanos/pull/5573
). Similar PR to #4118.
Arguments added:
* `--prometheus.get_config_timeout`
* `--prometheus.get_config_interval`

## Type of change

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

```release-note
Adds directives to configure Thanos arguments --prometheus.get_config_timeout --prometheus.get_config_interval
```

### CODE REVIEW

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 1]
```diff
@@ -668,6 +668,10 @@ type ThanosSpec struct {
 	BlockDuration Duration `json:"blockSize,omitempty"`
 	// ReadyTimeout is the maximum time Thanos sidecar will wait for Prometheus to start. Eg 10m
 	ReadyTimeout Duration `json:"readyTimeout,omitempty"`
+	// GetConfigInterval is how often the Prometheus config is fetched. Default 30s.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:671]: I wouldn't mention the default value as it's defined by Thanos and may change.

```suggestion
	// How often to retrieve the Prometheus configuration.
```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 2]
```diff
@@ -668,6 +668,10 @@ type ThanosSpec struct {
...
+	GetConfigInterval Duration `json:"getConfigInterval,omitempty"`
+	// GetConfigTimeout is the Timeout for getting Prometheus config. Default 5s.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:673]:
```suggestion
	// Maximum time to wait when retrieving the Prometheus configuration.
```

File: pkg/prometheus/server/statefulset.go [hunk 3]
```diff
@@ -730,6 +730,12 @@ func createThanosContainer(
 		if thanos.ReadyTimeout != "" {
 			thanosArgs = append(thanosArgs, monitoringv1.Argument{Name: "prometheus.ready_timeout", Value: string(thanos.ReadyTimeout)})
 		}
+		if thanos.GetConfigTimeout != "" {
```
Comment (simonpasquier) [inline pkg/prometheus/server/statefulset.go:733]: We need to check for `thanos.Version` as the flag requires v0.28.0 at least.

File: pkg/prometheus/server/statefulset.go [hunk 4]
```diff
@@ -730,6 +730,12 @@ func createThanosContainer(
...
+			thanosArgs = append(thanosArgs, monitoringv1.Argument{Name: "prometheus.get_config_timeout", Value: string(thanos.GetConfigTimeout)})
+		}
+		if thanos.GetConfigInterval != "" {
```
Comment (simonpasquier) [inline pkg/prometheus/server/statefulset.go:736]: same here

File: pkg/prometheus/server/statefulset.go (hunk 3 above)
Comment (viennaa) [reply pkg/prometheus/server/statefulset.go]: I've added the hashicorp version checker. If you have something specific in mind, I can also implement that instead.

Comment (simonpasquier) [reply pkg/prometheus/server/statefulset.go]: We already use `github.com/blang/semver/v4` for this. See for https://github.com/prometheus-operator/prometheus-operator/blob/dd19613873f75c402d64ec179433a3531a0cbe4a/pkg/prometheus/promcfg.go#L68-L72 an example

File: pkg/prometheus/server/statefulset.go [hunk 5]
```diff
@@ -731,6 +732,21 @@ func createThanosContainer(
...
 		}
 
+		thanosMinVersion, err := version.NewVersion("v0.28.0")
+		if err != nil {
+			return nil, err
+		}
+		thanosCurrentVersion, err := version.NewVersion(*thanos.Version)
+		if err != nil {
+			return nil, err
+		}
```
Comment (simonpasquier) [inline pkg/prometheus/server/statefulset.go:742]:
```suggestion
		thanosVersion, err := operator.StringPtrValOrDefault(thanos.Version, operator.DefaultThanosVersion)
		if err != nil {
			return nil, errors.Wrap(err, "failed to parse Prometheus version")
		}
```

File: pkg/prometheus/server/statefulset.go [hunk 6]
```diff
@@ -731,6 +732,21 @@ func createThanosContainer(
...
+		if thanos.GetConfigTimeout != "" && (thanosCurrentVersion.GreaterThanOrEqual(thanosMinVersion)) {
+			thanosArgs = append(thanosArgs, monitoringv1.Argument{Name: "prometheus.get_config_timeout", Value: string(thanos.GetConfigTimeout)})
+		}
+		if thanos.GetConfigInterval != "" && (thanosCurrentVersion.GreaterThanOrEqual(thanosMinVersion)) {
+			thanosArgs = append(thanosArgs, monitoringv1.Argument{Name: "prometheus.get_config_interval", Value: string(thanos.GetConfigInterval)})
+		}
```
Comment (simonpasquier) [inline pkg/prometheus/server/statefulset.go:748]:
```suggestion
		if thanos.GetConfigTimeout != "" && thanosVersion.GTE(semver.MustParse("0.18.0")) {
			thanosArgs = append(thanosArgs, monitoringv1.Argument{Name: "prometheus.get_config_timeout", Value: string(thanos.GetConfigTimeout)})
		}
		if thanos.GetConfigInterval != "" && thanosVersion.GTE(semver.MustParse("0.18.0")) {
			thanosArgs = append(thanosArgs, monitoringv1.Argument{Name: "prometheus.get_config_interval", Value: string(thanos.GetConfigInterval)})
		}
```

**Comment by viennaa** [comment]
Thanks for your input and fast response. I've used your semver code and adjusted the version + some minor things around. For sanity I just had another look when the [sidecar feature](https://github.com/thanos-io/thanos/pull/5573) went live and it [is available since v0.29.0 ](https://github.com/thanos-io/thanos/compare/v0.28.0...v0.29.0). 

Please let me know if I should rebase everything into one commit, or if your are fine with squashing in the PR.

**Comment by simonpasquier** [comment]
Indeed https://github.com/thanos-io/thanos/commit/06a64d699edeba33ce2340543a0717ab45529d46 shows it's available v0.29.0 while the CHANGELOG says v0.28.0...

**Review by simonpasquier — APPROVED** [review]
thanks!

**Pull request merged**